cipher.c

```diff
@@ -372,7 +372,7 @@ cipher_init(struct sshcipher_ctx **ccp, const struct sshcipher *cipher,
 		ret = SSH_ERR_ALLOC_FAIL;
 		goto out;
 	}
-	if (EVP_CipherInit(cc->evp, type, NULL, (u_char *)iv,
+	if (EVP_CipherInit(cc->evp, type, (u_char *)key, (u_char *)iv,
```

Comment:
I can't even remember why I did change (it's been more than a year). But I guess the same applies to the IV?

Reply:
Oh, I did make a comment about this:

    I've replaced the 2 EVP_CipherInit() calls in cipher_init() with 1. OpenSSL now clears
    everything when you call EVP_CipherInit() again, so what was passed in the first but not
    in the second call, and what the function calls between them did, was lost.

I wonder if this is actually a problem in OpenSSL.

Reply:
I think the IV can be passed in at either time? OpenSSL doesn't have a mechanism for ciphers with variable-length IVs as far as I know.

If the `EVP_CIPHER` has a fixed length key [and you already took care that what you're passing as the right size, since the `EVP_CipherInit` API isn't bounds-checked], you don't need to do `EVP_CIPHER_CTX_set_key_length` first. If you have a variable-length key cipher [and you're not using whatever the default key size is], you do.

Granted, it seems the only such ciphers, barring `ENGINE` mischief, are Blowfish, CAST, RC2, RC4, and RC5 so this pattern is only of so much interest... :-) Though the existing `EVP_CIPHER_CTX_set_key_length` logic did also act as a bounds check I think.
https://git.openssl.org/gitweb/?p=openssl.git&a=search&h=HEAD&st=grep&s=EVP_CIPH_VARIABLE_LENGTH
https://git.openssl.org/gitweb/?p=openssl.git&a=search&h=HEAD&st=grep&s=EVP_CIPH_CUSTOM_KEY_LENGTH

Reply:
Aha! I think what might actually be going on here is that openssh should have been using `EVP_CipherInit_ex`, not `EVP_CipherInit`. `EVP_CipherInit` was for folks who didn't call `EVP_CIPHER_CTX_init` (which `EVP_CIPHER_CTX_new` did implicitly).

But in 1.0.2, there was this weird quirk where `EVP_CipherInit` would skip `EVP_CIPHER_CTX_init` if `cipher` was NULL:
https://git.openssl.org/gitweb/?p=openssl.git;a=blob;f=crypto/evp/evp_enc.c;h=be577bac767fe9cf2c163be6967b20d1f6cc9014;hb=refs/heads/OpenSSL_1_0_2-stable#l94

OpenSSL 1.1.0 lost this behavior:
https://git.openssl.org/gitweb/?p=openssl.git;a=blob;f=crypto/evp/evp_enc.c;h=d21a0c7e54d54a5a77a2d5a6593e4851cf2f5291;hb=HEAD#l50

So I believe you want to revert this and merely switch those `EVP_CipherInit` calls to `EVP_CipherInit_ex`, which should work fine in 1.0.2 and 1.1.0. (Unless 1.1.0 additionally changed behavior here in ways I'm missing.) You may also wish to consider at minimum marking `EVP_CipherInit` as deprecated and documenting this behavior change, and possibly restoring the 1.0.2 behavior to avoid breaking folks.